### ISSUE TITLE
Make it easier to append/remove subnets

### ISSUE THREAD

**Issue by rosszurowski** [tailscale/tailscale#1436]
**Is your feature request related to a problem? Please describe.**

When adding a new subnet to a device, you need to enter the full list of subnets you want it to expose. However, if you don't already know which routes are exposed, it can be difficult to figure out. The only real way right now is to navigate to the admin panel to see which routes are listed, which can be a lengthy process or for some remote locations workflows, an impossibility.

For example, if my device exposes routes `a`, `b`, and `c` already, and I want to add `d`, how can I find out that `a`, `b`, and `c` are already exposed so I can:

```
sudo tailscale up --advertise-routes=a,b,c,d
```

**Describe the solution you'd like**

We could solve this two ways:
1. Add exposed subnet information to `tailscale status`, `tailscale status --json` or some other CLI location.
2. Allow users to append subnets with an `--append-route=` flag. This might also warrant a `--remove-route=` flag or a re-evaluation of how we want to handle routes on the CLI in general.
3. Something else?

**Comment by bradfitz** [comment]
It's not just routes. `tailscale up` doesn't let you modify anything else without specifying the exact set of everything. e.g. one customer had an outage because they wanted to toggle MagicDNS off so they did `tailscale up --accept-dns=false` (thinking it'd only affect things they named) and that turned off all their subnet routes, not knowing that `up` requires the full set of new state.

It all needs a rethink.

But, yes, we need some sort of "modify the current state" command. If we don't want to change the meaning of `up` for compatibility, maybe a new `set` subcommand that only mutates state, and we could remove `up` from docs and not list it from `--help` etc, so it's there but you have to know it's there to use it.

/cc @danderson

**Comment by rosszurowski** [comment]
Mm. A lot of command-line tools I've used separate a process like authenticating from later actions. Of course, they're very different tools than Tailscale, but maybe we can latch on to that same premise.

```bash
# log in once
tailscale login

# all other actions are stateful
tailscale routes add <route>
tailscale routes ls
tailscale dns disable
tailscale dns enable

# and reauthenticating doesn't require a hidden flag
tailscale login

# connect/disconnect are separate from auth, like on the GUIs
tailscale disconnect
tailscale connect
```

Of course, this approach would be annoying for admins who know exactly what they want and want the same config across many different devices. For them, the `set` approach would be more elegant.

**Comment by bradfitz** [comment]
When designing, you also need to consider how this fits into a multi-user world (either multiple simultaneous users or fast user switching).

**Comment by rosszurowski** [comment]
Yeah. Not saying this is the right approach, just a strawman proposal. Though the tools I'm referencing (Vercel, Heroku) allow for multiple "profiles" that are independent, and switching between them (eg. `vercel switch`). This will be more complex for us because we're interacting with a live `tailscaled` service, not some stateless REST API. But it seems conceivable that the same approach could work.

**Comment by DentonGentry** [comment]
Also see https://github.com/tailscale/tailscale/issues/2096

**Comment by kpfleming** [comment]
Along the same lines, I'd really like to be able to fully configure Tailscale (using Ansible, on my systems) without even bringing it up. If there was an option to manage the configuration in a plain text file (in whatever format you like) with the ability to validate a proposed configuration without applying it, automated management of Tailscale installations would be much easier than it is today.

In my specific case, I'm running Tailscale on a pair of HA (Pacemaker) nodes, where it will only ever be running on one of them at a time, but if I need to make a configuration change on the standby node I can't without bringing it up first. That effectively makes it a manual process.

**Comment by DentonGentry** [comment]
@kpfleming Could you open a new issue or comment on https://github.com/tailscale/tailscale/issues/270 instead?

**Comment by bradfitz** [comment]
Tailscale 1.34 included `tailscale set`. This bug remains open to add a way to modify part of a value (e.g. advertised routes or tags) with `tailscale set`.

Perhaps with a leading `+` or `-` to do a mutation:

```
$ tailscale set --advertised-routes=+10.0.0.0/8  # add 10.0.0.0/8 to whatever already exists
$ tailscale set --advertised-routes=-10.0.0.0/8  # remove 10.0.0.0/8 from whatever already exists
$ tailscale set --advertised-tags=+tag:foo # add tag:foo
```

/cc @DentonGentry @danderson @maisem

**Comment by spauka** [comment]
Apologies for the noise - since it looks like the last activity was 3 years ago, is this issue still live, or is there a workaround for modifying advertised routes without recreating the entire set command?
